phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexerRegionScanner.java

```diff
@@ -305,42 +320,96 @@ private void verifyIndex() throws IOException {
         verificationResult.add(nextVerificationResult);
     }
 
+    private void setMutationAttributes(Mutation m, byte[] uuidValue) {
+        m.setAttribute(useProto ? PhoenixIndexCodec.INDEX_PROTO_MD : PhoenixIndexCodec.INDEX_MD, indexMetaData);
+        m.setAttribute(PhoenixIndexCodec.INDEX_UUID, uuidValue);
+        m.setAttribute(BaseScannerRegionObserver.REPLAY_WRITES,
+                BaseScannerRegionObserver.REPLAY_INDEX_REBUILD_WRITES);
+        m.setAttribute(BaseScannerRegionObserver.CLIENT_VERSION, clientVersionBytes);
+        // Since we're replaying existing mutations, it makes no sense to write them to the wal
+        m.setDurability(Durability.SKIP_WAL);
+    }
+
+    private byte[] commitIfReady(byte[] uuidValue, UngroupedAggregateRegionObserver.MutationList mutationList) throws IOException {
+        if (ServerUtil.readyToCommit(mutationList.size(), mutationList.byteSize(), maxBatchSize, maxBatchSizeBytes)) {
+            ungroupedAggregateRegionObserver.checkForRegionClosing();
+            ungroupedAggregateRegionObserver.commitBatchWithRetries(region, mutationList, blockingMemstoreSize);
```

Comment:
Seems like it would be pretty straightforward to factor out the UARO.commitWith* functions into a dedicated Committer class that takes in a Region in the constructor and can be called from either UARO or here, to avoid the need for passing the UARO itself around. More in keeping with Single Responsibility Principle. 

Over time, we ideally should whittle UARO down to where it's just a dispatcher, catching Scan events and then calling appropriate objects to serve them based on the Scan attribute. That's out of scope here, but we can get a good start.

Reply:
Please see my related comments.